includes/class-wc-gateway-swedbank-pay-cc.php

```diff
@@ -657,36 +659,45 @@ public function thankyou_page( $order_id ) {
 			}
 		}
 
-		// Wait for order status update by a callback
-		$status = $order->get_status();
-		$times = 0;
-		$updated = false;
-		while ( true ) {
-			sleep( 1 );
-			clean_post_cache( $order->get_id() );
-			$order = wc_get_order( $order->get_id() );
-
-			// Check if status has been updated
-			if ( ! $order->has_status( $status ) ) {
-				$updated = true;
-				break;
-			}
-
-			// by timeout
-			$times ++;
-			if ( $times > 60 ) {
-				break;
-			}
-		}
+		if ( 'failed' === $order->get_status() ) {
+			// Wait for "Completed" transaction state
+			// Current payment can be changed
+			$attempts = 0;
+			while ( true ) {
+				sleep( 1 );
+				$attempts++;
+				if ($attempts > 60) {
+					break;
+				}
 
-		if ( ! $updated ) {
-			// Update an order status
-			try {
-				$this->core->fetchTransactionsAndUpdateOrder( $order_id );
-			} catch ( Exception $e ) {
-				$this->adapter->log(
-					LogLevel::WARNING, sprintf( 'fetchTransactionsAndUpdateOrder: %s', $e->getMessage() )
-				);
+				$transactions = $this->core->fetchTransactionsList( $payment_id );
+				foreach ($transactions as $transaction) {
+					if ( in_array( $transaction->getType(), ['Authorization', 'Sale'] ) ) {
+						switch ( $transaction->getState() ) {
+							case 'Completed':
+								// Transaction has found: update the order state
+								$this->core->fetchTransactionsAndUpdateOrder( $order_id, $transaction->getNumber() );
+
+								break 3;
+							case 'Failed':
+								// Log failed transaction
+								$this->adapter->log(
+									LogLevel::WARNING,
+									sprintf( 'Failed transaction: (%s), (%s), (%s), (%s), (%s), (%s), (%s)',
+										$order_id,
+										$payment_id,
+										$transaction->getId(),
+										$transaction->getData('failedReason'),
+										$transaction->getData('failedActivityName'),
+										$transaction->getData('failedErrorCode'),
+										$transaction->getData('failedErrorDescription')
```

Comment:
Don't use the `failedXxx` fields, [use the `problem` field with its RFC 7807 defined fields](https://developer.swedbankpay.com/payments/card/other-features#transaction-problems) instead.

Reply:
@asbjornu We don't show the problem to customer. We need to know the details of the problem and we log this

Reply:
Yes, but these fields are being deprecated. Use the ones found in `problem` instead.